server/src/main/java/org/elasticsearch/cluster/metadata/ProjectMetadata.java

```diff
@@ -1761,15 +1762,121 @@ static SortedMap<String, IndexAbstraction> buildIndicesLookup(
             if (indices.isEmpty()) {
                 return Collections.emptySortedMap();
             }
-            SortedMap<String, IndexAbstraction> indicesLookup = new TreeMap<>();
+            Map<String, IndexAbstraction> indicesLookup = new HashMap<>();
             Map<String, DataStream> indexToDataStreamLookup = new HashMap<>();
             collectDataStreams(dataStreamMetadata, indicesLookup, indexToDataStreamLookup);
 
             Map<String, List<IndexMetadata>> aliasToIndices = new HashMap<>();
             collectIndices(indices, indexToDataStreamLookup, indicesLookup, aliasToIndices);
             collectAliases(aliasToIndices, indicesLookup);
 
-            return Collections.unmodifiableSortedMap(indicesLookup);
+            // We do a ton of lookups on this map but also need its sorted properties at times.
+            // Using this hybrid of a sorted and a hash-map trades some heap overhead relative to just using a TreeMap
+            // for much faster O(1) lookups in large clusters.
+            return new SortedMap<>() {
+
+                private final SortedMap<String, IndexAbstraction> map = new TreeMap<>(indicesLookup);
```

Comment:
Can't we wrap the `TreeMap` in with `Collections.unmodifiableSortedMap()` on init already, instead of in every call?

Reply:
Right that makes sense, I started with just overriding the mutating methods to throw and only later realised that might not be enough and retrofitted the wrappers. Much cleaner this way, thanks!